serverless/serverless
# Added support for circular references in the variable system

## What did you implement:

Closes #3627 
Closes #3740
Closes #3711
Closes #3882

Added support for resolving circular dependencies in the variable system



## How did you implement it:
I'm keeping track of all the nodes that has been resolved along with their values into a map so that the variable system doesn't have to take the same walk twice (end of the circle)


## How can we verify it:
Try to reproduce issue #3627 , but to make sure nothing has been broken, use the following setup that includes a comprehensive coverage of the variable system edge cases, including the new circular support (`varQ` and `varR`).


- serverless.yml

```yml
service:
  name: s1

custom:
  data: # some data to reference
    zero: 0
    nooo: false
  varA: ${env:TESTING} # env vars
  varB: ${opt:stage} # opts vars
  varC: ${self:provider.name} # self vars
  varD: ${env:TESTING}-${opt:stage} # two vars in a string
  varE: ${opt:${env:TESTING}hoo} # nested vars
  varF: ${opt:region, opt:stage} # overwrite functionality
  varG: ${file(./vars.js):hello} # JS file running sync
  varH: ${file(./vars.js):promised} # JS file running async/promised!!!
  varI: ${file(./vars.json):hoo.hoo2} # deep JSON file
  varJ: ${file(./vars.yml):hee.hee2} # deep YAML file
  varK: my stage is ${opt:stage} # vars sub string
  varL: your account number is ${opt:number} # number vars as sub string
  varM: ${opt:number} # preserving data type
  varN: ${self:plugins, self:package, self:service} # multiple overwrites when empty object and undefined
  varO: ${self:custom.data.zero, self:service} # shouldn't overwrite 0 values
  varP: ${self:custom.data.nooo, self:service} # shouldn't overwrite false values
  varQ: ${self:} # referencing the entire config file, NEW!
  varR: ${self:custom.varQ.custom.varQ.provider.name} # circular, NEW!

provider:
  name: aws
  runtime: nodejs6.10

functions:
  f1:
    handler: handler.hello
```
- vars.js

```js
module.exports.hello = () => {
  // Sync code
  return {nested: 'world'};
}

module.exports.promised = () => {
  // Async code
  return Promise.resolve('world');
}
```

- vars.yml

```yml
hee:
  hee2: haa
```

- vars.json

```json
{
  "hoo": {
    "hoo2": "hum"
  }
}
```

- `export TESTING=ya`
- `sls package --stage dev --yahoo hi --region us-east-1 --number 1234`

After running this package command, you should see the resolved service in the console (I left out a console log for verification that I'll remove later)



## Todos:


***Is this ready for review?:*** YES
***Is it a breaking change?:*** NO
